src/main/java/org/spongepowered/common/event/tracking/phase/block/DispensePhaseState.java

```diff
@@ -58,8 +69,44 @@ public void unwind(final GeneralizedContext phaseContext) {
             });
         phaseContext.getCapturedEntitySupplier()
             .acceptAndClearIfNotEmpty(entities -> {
-                Sponge.getCauseStackManager().addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.DISPENSE);
-                SpongeCommonEventFactory.callSpawnEntity(entities, phaseContext);
+                List<Entity> projectiles = new ArrayList<>();
+                List<Entity> rest = new ArrayList<>();
+                for (Entity entity : entities) {
+                    if (entity instanceof Projectile) {
+                        projectiles.add(entity);
+                    } else {
+                        rest.add(entity);
+                    }
+                }
+
+                if (!rest.isEmpty()) {
+                    Sponge.getCauseStackManager().addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.DISPENSE);
+                    SpongeCommonEventFactory.callSpawnEntity(rest, phaseContext);
+                }
+
+                if (!projectiles.isEmpty()) {
+                    try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+                        frame.addContext(EventContextKeys.SPAWN_TYPE, SpawnTypes.PROJECTILE);
+                        TileEntity tileEntity = phaseContext.getSource(BlockSnapshot.class)
```

Comment:
Dispensing uses a `BlockSnapshot`? If you want to retain a reference to the original `TileEntity` and making this performant (we always do), make a specialized `PhaseContext` for `DispensePhaseState` that will take a `launcher(TileEntity)` and provide the nullable `TileEntity` that might be launching the projectile.